### PR TITLE
added onBeforeRequestAdd

### PR DESCRIPTION
**Pull request by kalimantos**
solves #145
- added `onBeforeRequestAdd(chip): Boolean`
- edited story for controlled input

### CODE REVIEW

File: src/ChipInput.js [hunk 1]
```diff
@@ -253,7 +253,7 @@ class ChipInput extends React.Component {
   }
 
   handleKeyDown = (event) => {
-    this.setState({keyPressed: false})
+    this.setState({keyPressed: false, preventChipCreation: false})
```
Comment (leMaik) [inline src/ChipInput.js]: Why do we need this state and handle the keyDown event? Only checking what the function says in `handleAddChip` should be sufficient.

File: src/ChipInput.js [hunk 2]
```diff
@@ -304,6 +304,7 @@ class ChipInput extends React.Component {
   }
 
   handleKeyUp = (event) => {
+    if (this.state.preventChipCreation) return
```
Comment (leMaik) [inline src/ChipInput.js]: Why is this needed?

File: src/ChipInput.js [hunk 3]
```diff
@@ -324,6 +325,9 @@ class ChipInput extends React.Component {
   }
 
   handleAddChip (chip) {
+    if (this.props.onBeforeRequestAdd && !this.props.onBeforeRequestAdd(chip)) {
+      return this.setState({ preventChipCreation })
```
Comment (leMaik) [inline src/ChipInput.js]: A simple `return` should be sufficient here. Also, where does the `preventChipCreation` variable come from here?

File: src/ChipInput.js (hunk 1 above)
Comment (kalimantos) [reply src/ChipInput.js]: it's all because of the `this.clearInput()` inside `handleKeyUp`
i need some more state to know when to prevent also the `handleKeyUp`

File: src/ChipInput.js (hunk 2 above)
Comment (kalimantos) [reply src/ChipInput.js]: this is needed to avoid input to be cleared, i tried removing it but this is the result:
bug-clear-input

**Comment by leMaik** [comment]
Looks fine to me now, @kalimantos. Is there anything you want to do on this before I merge the PR?

**Comment by jriuamb** [comment]
When the first chip is verified, the content of input it continues clearing

**Comment by leMaik** [comment]
@jriuamb I can't reproduce this. :open_mouth:

**Comment by jriuamb** [comment]
I am new to github. I'm trying to understand how it works. I downloaded the master branch version 0.18.1. I tried the Controlled Input form StoryBook. 
I have modified ControlledChipInput.js from stories folder, by setting the initial value of state with an empty array in the constructor. When you enter a value that onBeforeRequestAdd returns false, it does not clear the input.

`class ControlledChipInput extends React.Component {`
  `constructor (props) {`
  `super(props)`
    `this.state = {`
    `chips: []`
  `}`
`}`
`...`

Also happens that the console shows react errors.

image

**Comment by leMaik** [comment]
@jriuamb Thanks for the reproduction example! I'll look into this soon.
